application/src/main/java/org/thingsboard/server/service/edge/rpc/KafkaEdgeGrpcSession.java

```diff
@@ -0,0 +1,133 @@
+import org.thingsboard.server.queue.provider.TbCoreQueueFactory;
+import org.thingsboard.server.service.edge.EdgeContextComponent;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.function.BiConsumer;
+
+@Slf4j
+public class KafkaEdgeGrpcSession extends AbstractEdgeGrpcSession<KafkaEdgeGrpcSession> {
+
+    private final TbQueueConsumer<TbProtoQueueMsg<ToEdgeEventNotificationMsg>> edgeEventsConsumer;
+
+    private volatile boolean isHighPriorityProcessing;
+    private volatile boolean isConsumerInit;
+
+    private QueueConsumerManager<TbProtoQueueMsg<ToEdgeEventNotificationMsg>> consumer;
+
+    private ExecutorService consumerExecutor;
+
+    public KafkaEdgeGrpcSession(EdgeContextComponent ctx, TbCoreQueueFactory tbCoreQueueFactory, StreamObserver<ResponseMsg> outputStream,
+                                BiConsumer<EdgeId, KafkaEdgeGrpcSession> sessionOpenListener,
+                                BiConsumer<Edge, UUID> sessionCloseListener, ScheduledExecutorService sendDownlinkExecutorService,
+                                int maxInboundMessageSize, int maxHighPriorityQueueSizePerSession) {
+        super(ctx, outputStream, sessionOpenListener, sessionCloseListener, sendDownlinkExecutorService, maxInboundMessageSize, maxHighPriorityQueueSizePerSession);
+        edgeEventsConsumer = tbCoreQueueFactory.createEdgeEventMsgConsumer(tenantId, edge.getId());
+    }
+
+    private void processMsgs(List<TbProtoQueueMsg<ToEdgeEventNotificationMsg>> msgs, TbQueueConsumer<TbProtoQueueMsg<ToEdgeEventNotificationMsg>> consumer) {
+        log.trace("[{}][{}] starting processing edge events", tenantId, sessionId);
+        if (isConnected() && isSyncCompleted() && !isHighPriorityProcessing) {
+            List<EdgeEvent> edgeEvents = new ArrayList<>();
+            for (TbProtoQueueMsg<ToEdgeEventNotificationMsg> msg : msgs) {
+                EdgeEvent edgeEvent = ProtoUtils.fromProto(msg.getValue().getEdgeEventMsg());
+                edgeEvents.add(edgeEvent);
+            }
+            List<DownlinkMsg> downlinkMsgsPack = convertToDownlinkMsgsPack(edgeEvents);
+            try {
+                boolean isInterrupted = sendDownlinkMsgsPack(downlinkMsgsPack).get();
+                if (isInterrupted) {
+                    log.debug("[{}][{}][{}] Send downlink messages task was interrupted", tenantId, edge.getId(), sessionId);
+                } else {
+                    consumer.commit();
+                }
+            } catch (Exception e) {
+                log.error("[{}] Failed to process all downlink messages", sessionId, e);
+            }
+        } else {
+            log.trace("[{}][{}] edge is not connected or sync is not completed. Skipping iteration", tenantId, sessionId);
```

Comment:
When you don't commit, on the next poll Kafka will **immediately** return the same msg pack. I suggest that you add some sleep here.